system/jlib/jmutex.cpp

```diff
@@ -329,6 +321,53 @@ void Monitor::notifyAll()
 
 //==================================================================================
 
+#ifdef USE_PTHREAD_RWLOCK
+
+bool ReadWriteLock::lockRead(unsigned timeout)
+{
+    if (timeout == (unsigned)-1)
+    {
+        lockRead();
+        return true;
+    }
+
+    if (pthread_rwlock_tryrdlock(&rwlock) == 0)
+        return true;
+
+    timespec endtime;
+    getEndTime(endtime, timeout);
+    return (pthread_rwlock_timedrdlock(&rwlock, &endtime) == 0);
+}
+
+bool ReadWriteLock::lockWrite(unsigned timeout)
+{
+    if (timeout == (unsigned)-1)
+    {
+        lockWrite();
+        return true;
+    }
+
+    if (pthread_rwlock_trywrlock(&rwlock) == 0)
+        return true;
+
+    timespec endtime;
+    getEndTime(endtime, timeout);
+    return (pthread_rwlock_timedwrlock(&rwlock, &endtime) == 0);
+}
+
+bool ReadWriteLock::queryWriteLocked()
```

Comment:
I agree - I don't like this function, but I couldn't think of any other way of implementing it.  Ideas welcomed.

Actually the only calls to it could be deleted, the code could call unlock() instead.  I will release a separate fix.

Reply:
I take that back, it uses the value in the block() call().  I suspect it could be recoded without it - by passing it into the class, but not trivially.

Reply:
Why/where query a lock without locking it ?

Reply:
It is called from some code that checks if was write locked before unlocking it.  I think I'll delete it from the PR (so the new code could not be used as a direct replacement).